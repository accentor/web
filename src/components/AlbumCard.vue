<template>
  <VCard :to="{ name: 'album', params: { id: album.id } }">
    <VImg
      :aspect-ratio="1"
      :src="album.image500"
      v-if="album.image500 && !imageUnavailable"
      @error="setImageUnavailable"
    />
    <VImg
      :aspect-ratio="1"
      :src="album.image"
      v-else-if="album.image && !imageUnavailable"
      @error="setImageUnavailable"
    />
    <VImg
      :aspect-ratio="1"
      :src="require('@mdi/svg/svg/album.svg')"
      v-else
      class="grey lighten-3"
    />
    <VCardTitle class="pb-0 d-block text-truncate" :title="full_title">
      {{ album.title }}&nbsp;
      <span
        v-if="
          album.edition_description !== null && album.edition_description.length
        "
        class="grey--text"
      >
        ({{ album.edition_description }})
      </span>
    </VCardTitle>
    <VCardText>
      <AlbumArtists :album="album" :truncate="true" />
    </VCardText>
    <VCardText>
      <div class="grey--text">{{ album.release }}</div>
      <div v-if="labelForCatNr" class="grey--text">
        {{ catalogueNumber || $t("music.label.catalogue-number-none") }}
      </div>
    </VCardText>
    <VCardActions>
      <AlbumActions :album="album" class="actions--sm-wide" />
    </VCardActions>
  </VCard>
</template>
<script>
import AlbumActions from "./AlbumActions";
import AlbumArtists from "./AlbumArtists";

export default {
  name: "AlbumCard",
  components: { AlbumArtists, AlbumActions },
  props: {
    album: {
      type: Object,
      required: true,
    },
    labelForCatNr: {
      type: Object,
      required: false,
    },
  },
  data() {
    return {
      imageUnavailable: false,
    };
  },
  computed: {
    catalogueNumber() {
      if (this.labelForCatNr) {
        return this.album.album_labels.find(
          (al) => al.label_id === this.labelForCatNr.id
        ).catalogue_number;
      } else {
        return undefined;
      }
    },
    full_title() {
      let full_title = this.album.title;
      if (this.album.edition_description) {
        full_title += ` ${this.album.edition_description}`;
      }
      return full_title;
    },
  },
  methods: {
    setImageUnavailable() {
      this.imageUnavailable = true;
    },
  },
};
</script>
