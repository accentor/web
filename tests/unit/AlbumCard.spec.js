import Vuetify from "vuetify";
import VueRouter from "vue-router";
import { mount, createLocalVue } from "@vue/test-utils";
import AlbumCard from "@/components/AlbumCard.vue";

const localVue = createLocalVue();
localVue.use(VueRouter);

const factory = (values = {}, album_values = {}) => {
  const vuetify = new Vuetify();
  const router = new VueRouter();

  return mount(AlbumCard, {
    localVue,
    vuetify,
    router,
    propsData: {
      album: {
        id: 1,
        title: "title",
        album_artists: [],
        album_labels: [
          {
            label_id: 1,
            catalogue_number: "catNr",
          },
          {
            label_id: 2,
            catalogue_number: "otherCatNr",
          },
        ],
        edition: null,
        edition_description: null,
        release: "1970-01-01",
        ...album_values,
      },
      ...values,
    },
    mocks: {
      $t: (msg) => msg,
    },
    stubs: ["AlbumActions"],
  });
};

describe("AlbumCard", () => {
  const wrapper = factory();

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders album title", () => {
    expect(wrapper.find(".v-card__title").text()).toMatch(/title/);
  });

  it("does not renders album title with edition description", () => {
    expect(wrapper.find(".v-card__title span")).toBeUndefined;
  });

  it("renders release date", () => {
    expect(wrapper.find(".album-release").text()).toMatch("1970-01-01");
  });

  it("does not render catalogue number", () => {
    expect(wrapper.find(".album-cat")).toBeUndefined;
  });
});

describe("AlbumCard with release description", () => {
  const wrapper = factory({}, { edition_description: "description" });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders album title with edition description", () => {
    expect(wrapper.find(".v-card__title span").text()).toMatch("(description)");
  });
});

describe("AlbumCard with CatNr", () => {
  const wrapper = factory({ labelForCatNr: { id: 1 } });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders correct catalogue number", () => {
    expect(wrapper.find(".album-cat").text()).toMatch("catNr");
  });
});
