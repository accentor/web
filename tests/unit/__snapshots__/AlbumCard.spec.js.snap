// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumCard renders correctly 1`] = `
<a
  aria-current="page"
  class="router-link-exact-active router-link-active v-card v-card--link v-sheet theme--light"
  href="#/"
  tabindex="0"
>
  <div
    class="v-image v-responsive grey lighten-3 theme--light"
  >
    <div
      class="v-responsive__sizer"
      style="padding-bottom: 100%;"
    />
    <transition-stub
      mode="in-out"
      name="fade-transition"
    >
      <div
        class="v-image__image v-image__image--preload v-image__image--cover"
        style="background-position: center center;"
      />
    </transition-stub>
    <div
      class="v-responsive__content"
    />
  </div>
   
  <div
    class="v-card__title pb-0 d-block text-truncate"
    title="title"
  >
    
    title 
    
    <!---->
  </div>
   
  <div
    class="v-card__text"
  >
    <div>
      <span>
        music.artist.various
      </span>
    </div>
  </div>
   
  <div
    class="v-card__text"
  >
    <div
      class="grey--text album-release"
    >
      1970-01-01
    </div>
     
    <!---->
  </div>
   
  <div
    class="v-card__actions"
  >
    <albumactions-stub
      album="[object Object]"
    />
  </div>
</a>
`;

exports[`AlbumCard with CatNr renders correctly 1`] = `
<a
  aria-current="page"
  class="router-link-exact-active router-link-active v-card v-card--link v-sheet theme--light"
  href="#/"
  tabindex="0"
>
  <div
    class="v-image v-responsive grey lighten-3 theme--light"
  >
    <div
      class="v-responsive__sizer"
      style="padding-bottom: 100%;"
    />
    <transition-stub
      mode="in-out"
      name="fade-transition"
    >
      <div
        class="v-image__image v-image__image--preload v-image__image--cover"
        style="background-position: center center;"
      />
    </transition-stub>
    <div
      class="v-responsive__content"
    />
  </div>
   
  <div
    class="v-card__title pb-0 d-block text-truncate"
    title="title"
  >
    
    title 
    
    <!---->
  </div>
   
  <div
    class="v-card__text"
  >
    <div>
      <span>
        music.artist.various
      </span>
    </div>
  </div>
   
  <div
    class="v-card__text"
  >
    <div
      class="grey--text album-release"
    >
      1970-01-01
    </div>
     
    <div
      class="grey--text album-cat"
    >
      
      catNr
    
    </div>
  </div>
   
  <div
    class="v-card__actions"
  >
    <albumactions-stub
      album="[object Object]"
    />
  </div>
</a>
`;

exports[`AlbumCard with release description renders correctly 1`] = `
<a
  aria-current="page"
  class="router-link-exact-active router-link-active v-card v-card--link v-sheet theme--light"
  href="#/"
  tabindex="0"
>
  <div
    class="v-image v-responsive grey lighten-3 theme--light"
  >
    <div
      class="v-responsive__sizer"
      style="padding-bottom: 100%;"
    />
    <transition-stub
      mode="in-out"
      name="fade-transition"
    >
      <div
        class="v-image__image v-image__image--preload v-image__image--cover"
        style="background-position: center center;"
      />
    </transition-stub>
    <div
      class="v-responsive__content"
    />
  </div>
   
  <div
    class="v-card__title pb-0 d-block text-truncate"
    title="title description"
  >
    
    title 
    
    <span
      class="grey--text"
    >
      
      (description)
    
    </span>
  </div>
   
  <div
    class="v-card__text"
  >
    <div>
      <span>
        music.artist.various
      </span>
    </div>
  </div>
   
  <div
    class="v-card__text"
  >
    <div
      class="grey--text album-release"
    >
      1970-01-01
    </div>
     
    <!---->
  </div>
   
  <div
    class="v-card__actions"
  >
    <albumactions-stub
      album="[object Object]"
    />
  </div>
</a>
`;
